### PR TITLE
ARROW-15277: [C++][Python] Use ChunkedArray::Make for chunked_array

### PR DESCRIPTION
**Pull request by milesgranger**
Supersedes and will close #12096 

[ARROW-15277](https://issues.apache.org/jira/browse/ARROW-15277)

### CODE REVIEW

**Comment by milesgranger** [comment]
@pitrou I believe this is ready for a review.

File: python/pyarrow/tests/test_table.py [hunk 1]
```diff
@@ -143,14 +140,15 @@ def test_chunked_array_to_numpy():
 
 
 def test_chunked_array_mismatch_types():
-    with pytest.raises(TypeError):
+    msg = "chunks must all be same type"
+    with pytest.raises(pa.ArrowInvalid, match=msg):
```
Comment (pitrou) [inline python/pyarrow/tests/test_table.py:144]: Hmm... interesting. I would be in favor of changing the C++ side to return `TypeError` instead. Would you like to do that in this PR?

File: python/pyarrow/tests/test_table.py [hunk 2]
```diff
@@ -143,14 +140,15 @@ def test_chunked_array_to_numpy():
...
         # Given array types are different
         pa.chunked_array([
             pa.array([1, 2, 3]),
             pa.array([1., 2., 3.])
         ])
 
-    with pytest.raises(TypeError):
+    with pytest.raises(pa.ArrowInvalid, match=msg):
```
Comment (pitrou) [inline python/pyarrow/tests/test_table.py:151]: (same)

**Comment by milesgranger** [comment]
I don't think the [R / AMD64 Windows R 3.6 RTools 35 failure](https://github.com/apache/arrow/runs/8003655286?check_suite_focus=true) is related...

**Comment by pitrou** [comment]
Indeed, it isn't.

**Comment by pitrou** [comment]
Rebased and a nit.

**Review by pitrou — APPROVED** [review]
Thanks @milesgranger !

**Comment by pitrou** [comment]
Python CI was successful here: https://github.com/milesgranger/arrow/actions/runs/2924730791

**Pull request merged**